File: backend/app/schemas/metrics.py
"""Schemas for /health and /metrics endpoints."""
from __future__ import annotations

from typing import Dict, List, Optional

from pydantic import BaseModel, Field


class HealthOut(BaseModel):
    """GET /health response."""

    status: str = Field(..., description="Either 'ok' or 'degraded'.")
    version: str = Field(..., description="Backend app version.")
    artifacts_loaded_at: Optional[str] = Field(
        default=None,
        description="ISO-8601 timestamp when artifacts were loaded. Null if not loaded.",
    )
    item_count: int = Field(default=0, ge=0, description="Number of items in artifacts.")
    context_count: int = Field(default=0, ge=0)
    embedding_dim: int = Field(default=0, ge=0)
    research_mode: bool = False
    embedding_backend: str = Field(default="not_loaded")
    e5_model: str = Field(default="")
    e5_load_latency_ms: Optional[float] = Field(default=None, ge=0)
    e5_last_inference_latency_ms: Optional[float] = Field(default=None, ge=0)
    e5_error: Optional[str] = None


class MetricsOut(BaseModel):
    """GET /metrics response — high-level corpus stats."""

    item_count: int = Field(..., ge=0)
    context_count: int = Field(..., ge=0)
    keyword_count: int = Field(..., ge=0)
    positive_user_count: int = Field(..., ge=0)
    unique_item_user_edges: int = Field(..., ge=0)
    embedding_dim: int = Field(..., ge=0)
    artifacts_loaded_at: str = Field(..., description="ISO-8601 timestamp.")
    config_hash: str = Field(..., description="Short hash of the pipeline config used.")


class RequestTrendBucket(BaseModel):
    """One month-bucket in the dashboard trend chart."""

    year: int = Field(..., ge=1970, le=2100)
    month: int = Field(..., ge=1, le=12, description="1-based calendar month.")
    label: str = Field(..., description="Short Thai month label, e.g. 'ส.ค.'")
    request_count: int = Field(..., ge=0, description="Recommendations requested.")
    shown_count: int = Field(
        ...,
        ge=0,
        description="Items actually shown (sum of result rows in those requests).",
    )


class RequestTrendOut(BaseModel):
    """GET /metrics/requests response — dashboard trend chart."""

    months: int = Field(..., ge=1, le=36, description="Window size (last N months).")
    total_requests: int = Field(..., ge=0)
    total_shown: int = Field(..., ge=0)
    source: str = Field(
        ...,
        description="'postgres' when DB is reachable, 'disabled' otherwise.",
    )
    buckets: List[RequestTrendBucket] = Field(
        default_factory=list,
        description="Oldest first; length == months.",
    )


class ModelConfigOut(BaseModel):
    """GET /metrics/config response — chosen experiment config the
    backend currently serves. Mirrors ``artifacts/outputs/best_model_config.json``.
    Falls back to env-var defaults when the manifest is absent."""

    cbf_model: str = Field(default="", description="CBF encoder model id.")
    cf_model: str = Field(default="", description="CF model class.")
    hybrid_method: str = Field(default="", description="Hybrid combiner method.")
    hybrid_alpha: Optional[float] = Field(
        default=None,
        description="CBF weight in the WeightedSum combiner. CF weight = 1 - alpha.",
    )
    candidate_strategy: str = Field(default="", description="Eligibility filter name.")
    embedding_dim: Optional[int] = Field(
        default=None,
        description="Embedding dimensionality served by the loader.",
    )
    itemknn_k: Optional[int] = Field(
        default=None,
        description="ItemKNN top-K neighbour count.",
    )
    itemknn_shrink: Optional[float] = Field(
        default=None,
        description="Cosine shrinkage constant.",
    )
    cbf_keyword_boost: Optional[float] = Field(
        default=None,
        description="Additive boost on keyword match in CBF scoring.",
    )
    positive_threshold: Optional[int] = Field(
        default=None,
        description="Min rating to count as a positive CF signal.",
    )
    extra: Dict[str, object] = Field(
        default_factory=dict,
        description="Other keys from best_model_config.json surfaced as-is.",
    )


class ReproducibilityCount(BaseModel):
    expected: int = Field(..., ge=0)
    actual: int = Field(..., ge=0)
    delta: int
    matches: bool


class ReproducibilityOut(BaseModel):
    """Paper baseline compared with the currently loaded artifact corpus."""

    status: str = Field(..., description="'match' or 'drift_detected'.")
    baseline_source: str
    keyword_count_source: str = Field(description="'postgres' or 'artifacts'.")
    item_count: ReproducibilityCount
    keyword_count: ReproducibilityCount
    taxonomy_path_count: ReproducibilityCount
    artifact_config_hash: str = ""
    artifact_build_timestamp: str = ""

File: frontend/lib/server/metrics.ts
// lib/server/metrics.ts — corpus metrics, recommendation-request trend, and
// the active recommender configuration, served by the Application Backend.
//
// These are the issue #9 ports of the legacy FastAPI ``/metrics``,
// ``/metrics/requests`` and ``/metrics/config`` routes. The response shapes
// are unchanged so existing consumers (homepage, researcher tooling) keep
// working; only the data sources moved: live counts come from the TypeORM
// tables instead of the legacy SQLAlchemy session, and the artifact-derived
// fields (CF index stats, embedding dim) are zeroed because the Application
// Backend never opens ``artifacts/`` (architecture invariant #2).

import type { MetricsOut, ModelConfigOut, RequestTrendBucket, RequestTrendOut } from "@/lib/types";
import { getDataSource } from "@/db/connection";
import { CatalogueContextEntity, CatalogueItemEntity, CatalogueKeywordEntity } from "@/db/entities/Catalogue";
import { RecommendationRequestEntity, RecommendationResultEntity } from "@/db/entities/RecommendationRequests";

const THAI_MONTH_LABELS = [
  "", "ม.ค.", "ก.พ.", "มี.ค.", "เม.ย.", "พ.ค.", "มิ.ย.",
  "ก.ค.", "ส.ค.", "ก.ย.", "ต.ค.", "พ.ย.", "ธ.ค.",
];

function monthWindow(months: number): { startYear: number; startMonth: number } {
  const today = new Date();
  let year = today.getUTCFullYear();
  let month = today.getUTCMonth() + 1;
  for (let step = 1; step < months; step += 1) {
    month -= 1;
    if (month === 0) {
      month = 12;
      year -= 1;
    }
  }
  return { startYear: year, startMonth: month };
}

function* iterateMonthBuckets(
  startYear: number,
  startMonth: number,
  count: number,
): Generator<{ year: number; month: number; label: string }> {
  let year = startYear;
  let month = startMonth;
  for (let index = 0; index < count; index += 1) {
    yield { year, month, label: THAI_MONTH_LABELS[month] };
    month += 1;
    if (month === 13) {
      month = 1;
      year += 1;
    }
  }
}

/**
 * ``GET /metrics`` — aggregated corpus counts plus the build identity.
 * Item/context/keyword counts come from the live tables; the CF-index
 * fields (positive users, edges, embedding dim, artifact build) are
 * artifact-owned and therefore reported as zeros/empty by the Application
 * Backend (the private model service holds that state).
 */
export async function corpusMetrics(): Promise<MetricsOut> {
  const dataSource = await getDataSource();
  const [itemCount, contextCount, keywordCount] = await Promise.all([
    dataSource.getRepository(CatalogueItemEntity).countBy({ isActive: true }),
    dataSource.getRepository(CatalogueContextEntity).count(),
    dataSource.getRepository(CatalogueKeywordEntity).count(),
  ]);
  return {
    item_count: itemCount,
    context_count: contextCount,
    keyword_count: keywordCount,
    positive_user_count: 0,
    unique_item_user_edges: 0,
    embedding_dim: 0,
    artifacts_loaded_at: "",
    config_hash: "",
  };
}

/**
 * ``GET /metrics/requests?months=N`` — monthly recommendation-request
 * trend aggregated from the persisted ``recommendation_requests`` /
 * ``recommendation_results`` rows (the same tables the legacy dashboard
 * chart read). Returns zeroed buckets with ``source: "disabled"`` when
 * the database is unavailable so the chart never 500s.
 */
// fallow-ignore-next-line complexity -- Ported 1:1 from the legacy analytics service; splitting would break parity with the reference implementation.
export async function requestTrend(months = 12): Promise<RequestTrendOut> {
  const safeMonths = Math.min(36, Math.max(1, Math.floor(months)));
  const { startYear, startMonth } = monthWindow(safeMonths);
  const bucketsIndex = new Map<string, { requestCount: number; shownCount: number }>();
  for (const bucket of iterateMonthBuckets(startYear, startMonth, safeMonths)) {
    bucketsIndex.set(`${bucket.year}:${bucket.month}`, { requestCount: 0, shownCount: 0 });
  }

  let source: RequestTrendOut["source"] = "disabled";
  let totalRequests = 0;
  let totalShown = 0;
  try {
    const dataSource = await getDataSource();
    const since = new Date(Date.UTC(startYear, startMonth - 1, 1));
    const requests = await dataSource
      .getRepository(RecommendationRequestEntity)
      .createQueryBuilder("request")
      .select('EXTRACT(YEAR FROM request.created_at)', "y")
      .addSelect('EXTRACT(MONTH FROM request.created_at)', "m")
      .addSelect("COUNT(*)", "c")
      .where("request.created_at >= :since", { since })
      .groupBy("y")
      .addGroupBy("m")
      .getRawMany<{ y: string; m: string; c: string }>();
    const shown = await dataSource
      .getRepository(RecommendationResultEntity)
      .createQueryBuilder("result")
      .innerJoin("RecommendationRequest", "request", "request.id = result.request_id")
      .select('EXTRACT(YEAR FROM request.created_at)', "y")
      .addSelect('EXTRACT(MONTH FROM request.created_at)', "m")
      .addSelect("COUNT(*)", "c")
      .where("request.created_at >= :since", { since })
      .groupBy("y")
      .addGroupBy("m")
      .getRawMany<{ y: string; m: string; c: string }>();
    source = "postgres";
    for (const row of requests) {
      const key = `${Number(row.y)}:${Number(row.m)}`;
      const bucket = bucketsIndex.get(key);
      if (bucket) {
        bucket.requestCount += Number(row.c);
        totalRequests += Number(row.c);
      }
    }
    for (const row of shown) {
      const key = `${Number(row.y)}:${Number(row.m)}`;
      const bucket = bucketsIndex.get(key);
      if (bucket) {
        bucket.shownCount += Number(row.c);
        totalShown += Number(row.c);
      }
    }
  } catch {
    // Dashboard trend must degrade to zeros, never 500.
    source = "disabled";
    totalRequests = 0;
    totalShown = 0;
  }

  const buckets: RequestTrendBucket[] = [];
  for (const { year, month, label } of iterateMonthBuckets(startYear, startMonth, safeMonths)) {
    const data = bucketsIndex.get(`${year}:${month}`) ?? { requestCount: 0, shownCount: 0 };
    buckets.push({
      year,
      month,
      label,
      request_count: data.requestCount,
      shown_count: data.shownCount,
    });
  }
  return {
    months: safeMonths,
    total_requests: totalRequests,
    total_shown: totalShown,
    source,
    buckets,
  };
}

function envFloat(name: string, fallback: number): number {
  const raw = process.env[name];
  const parsed = Number(raw);
  return raw !== undefined && raw !== "" && Number.isFinite(parsed) ? parsed : fallback;
}

function envInt(name: string, fallback: number): number {
  const raw = process.env[name];
  const parsed = Number(raw);
  return raw !== undefined && raw !== "" && Number.isSafeInteger(parsed) ? parsed : fallback;
}

/**
 * ``GET /metrics/config`` — the active recommender configuration the
 * Application Backend serves. Mirrors the legacy response built from
 * ``best_model_config.json`` + ``RECSYS_*`` env overrides; here the
 * values are derived from the same ``RECSYS_*`` environment so the
 * dashboard's model sliders keep reflecting reality.
 */
export function activeModelConfig(): ModelConfigOut {
  return {
    cbf_model: process.env.RECSYS_E5_MODEL || "intfloat/multilingual-e5-large-instruct",
    cf_model: "ItemKNN",
    hybrid_method: "Hybrid-WeightedSum",
    hybrid_alpha: envFloat("RECSYS_HYBRID_ALPHA", 0.7),
    candidate_strategy: "EligibilityGate",
    embedding_dim: null,
    itemknn_k: envInt("RECSYS_ITEMKNN_K", 10),
    itemknn_shrink: envFloat("RECSYS_ITEMKNN_SHRINK", 50.0),
    cbf_keyword_boost: envFloat("RECSYS_CBF_KEYWORD_BOOST", 0.05),
    positive_threshold: envInt("RECSYS_POSITIVE_THRESHOLD", 4),
    extra: {},
  };
}
